ibverbs/src/lib.rs

```diff
@@ -377,31 +365,24 @@ impl Context {
                 ));
             }
         }
+        Ok(port_attr)
+    }
 
-        let mut gid_table = vec![ffi::ibv_gid_entry::default(); port_attr.gid_tbl_len as usize];
-        let num_entries = unsafe {
-            ffi::_ibv_query_gid_table(
-                ctx,
-                gid_table.as_mut_ptr(),
-                gid_table.len(),
-                0,
-                size_of::<ffi::ibv_gid_entry>(),
-            )
-        };
-        if num_entries < 0 {
+    /// Opens a context for the given device, and queries its port and gid.
+    fn with_device(dev: *mut ffi::ibv_device) -> io::Result<Context> {
+        assert!(!dev.is_null());
+
+        let ctx = unsafe { ffi::ibv_open_device(dev) };
+        if ctx.is_null() {
             return Err(io::Error::new(
                 io::ErrorKind::Other,
-                format!("failed to query gid table, error={}", -num_entries),
+                "failed to open device".to_string(),
             ));
         }
-        gid_table.truncate(num_entries as usize);
-        let gid_table = gid_table.into_iter().map(GidEntry::from).collect();
 
-        Ok(Context {
-            ctx,
-            port_attr,
-            gid_table,
-        })
+        let ctx = Context { ctx };
+        ctx.query_port()?;
```

Comment:
So we don't need to store the result of this call any more because it actually needs to be queried per QP, not just once for the whole device context? In that case, why even have the call here at all?